contributors/design-proposals/storage/csi-migration.md

```diff
@@ -270,9 +317,11 @@ Kubelet. The workflow is as follows:
 2. A/D Controller restarted and migrated (flags flipped)
 3. A/D Controller continues to use in-tree code for this node b/c node
    annotation doesn't exist
-4. Node drained and made unschedulable. All volumes unmounted/detached with in-tree code
+4. Node drained and made unschedulable. All volumes unmounted/detached with
+   in-tree code
 5. Kubelet restarted and migrated (flags flipped)
-6. Kubelet annotates node to tell A/D controller this node has been migrated
+6. Kubelet updates CSINodeInfo node to tell A/D controller whether each
+   node/driver has been migrated
 7. Kubelet is made schedulable
 8. Both A/D Controller & Kubelet Migrated, node is in "fresh" state so all new
```

Comment:
6 and 8 sound racy to me, scheduler can get Node update before A/D controller gets CSINodeInfo from step 6. Remember not to use informer for CSINodeInfo. Explicit node somewhere would be nice so we don't forget this.

Reply:
Is this racy only when using informers? The Kubelet should be posting `NotReady` state until CSINodeInfo's have all been updated so from that perspective I don't think scheduler will schedule to the node until the fields required by ADC are propagated. Then as long as there's no informer lag (maybe this is not an assumption I can make) there should be no race right?

Reply:
>  I don't think scheduler will schedule to the node until the fields required by ADC are propagated.

That's correct, however, it will work only when A/D controller and scheduler have the same shared informer. IMO we still run scheduler as a separate process and its informer can get watch events at different pace / in different order than A/D controller.

Reply:
I see, discussed with @msau42 and it seems like the best solution we can come up with for now is to not use informer on attach to guarantee up-to-date `CSINodeInfo`. It will be important to test and document performance impact of this.